src/pk.c

```diff
@@ -6959,793 +7110,1199 @@ WOLFSSL_BIGNUM* wolfSSL_DH_8192_prime(WOLFSSL_BIGNUM* bn)
+        }
     }
+}
 
-    if (wolfSSL_BIO_set_fp(fbio, fp, BIO_NOCLOSE) != 1) {
-        wolfSSL_BIO_free(fbio);
-        WOLFSSL_MSG("wolfSSL_BIO_set_fp error");
-        return NULL;
+/* Set the public and/or private key.
+ *
+ * @param [in, out] dh        DH key to have keys set into.
+ * @param [in]      pub_key   Public key to set. May be NULL.
+ * @param [in]      priv_key  Private key to set. May be NULL.
+ * @return  0 on failure.
+ * @return  1 on success.
+ */
+int wolfSSL_DH_set0_key(WOLFSSL_DH *dh, WOLFSSL_BIGNUM *pub_key,
+    WOLFSSL_BIGNUM *priv_key)
+{
+    int ret = 1;
+
+    WOLFSSL_ENTER("wolfSSL_DH_set0_key");
+
+    /* Validate parameters. */
+    if (dh == NULL) {
+        ret = 0;
     }
 
-    /* wolfSSL_PEM_read_bio_DHparams() sanitizes x, cb, u args */
-    dh = wolfSSL_PEM_read_bio_DHparams(fbio, x, cb, u);
-    wolfSSL_BIO_free(fbio);
-    return dh;
+    /* Replace public key when one passed in. */
+    if ((ret == 1) && (pub_key != NULL)) {
+        wolfSSL_BN_free(dh->pub_key);
+        dh->pub_key = pub_key;
+    }
+
+    /* Replace private key when one passed in. */
+    if ((ret == 1) && (priv_key != NULL)) {
+        wolfSSL_BN_free(dh->priv_key);
+        dh->priv_key = priv_key;
+    }
+
+    /* Update internal key if parameters set. */
+    if ((ret == 1) && ((dh->p != NULL) && (dh->g != NULL))) {
+        ret = SetDhInternal(dh);
+    }
+
+    return ret;
 }
-#endif /* !NO_FILESYSTEM */
 
-#endif /* !NO_BIO */
+#endif /* OPENSSL_EXTRA */
 
-#if defined(WOLFSSL_DH_EXTRA) && !defined(NO_FILESYSTEM)
-/* Writes the DH parameters in PEM format from "dh" out to the file pointer
- * passed in.
+/*
+ * DH check APIs
+ */
+
+#ifdef OPENSSL_EXTRA
+
+#ifndef NO_CERTS
+
+#ifdef OPENSSL_ALL
+/* Check whether BN number is a prime.
+ *
+ * @param [in]  n        Number to check.
+ * @param [out] isPrime  MP_YES when prime and MP_NO when not.
+ * @return  1 on success.
+ * @return  0 on error.
+ */
+static int wolfssl_dh_check_prime(WOLFSSL_BIGNUM* n, int* isPrime)
+{
+    int ret = 1;
+#ifdef WOLFSSL_SMALL_STACK
+    WC_RNG* tmpRng = NULL;
+#else
+    WC_RNG  tmpRng[1];
+#endif
+    WC_RNG* rng;
+    int localRng;
+
+    /* Make an RNG with tmpRng or get global. */
+    rng = wolfssl_make_rng(tmpRng, &localRng);
+    if (rng == NULL) {
+        ret = 0;
+    }
+    if (ret == 1) {
+        mp_int* prime = (mp_int*)n->internal;
+
+        if (mp_prime_is_prime_ex(prime, 8, isPrime, rng) != 0) {
+            ret = 0;
+        }
+        /* Free local random number generator if created. */
+        if (localRng) {
+            wc_FreeRng(rng);
+        #ifdef WOLFSSL_SMALL_STACK
+            XFREE(rng, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+        #endif
+        }
+    }
+
+    return ret;
+}
+
+/* Checks the Diffie-Hellman parameters.
+ *
+ * Checks that the generator and prime are available.
+ * Checks that the prime is prime.
+ * OpenSSL does not return 0 when no operational failures buf error codes set.
```

Comment:
Change to "but error codes will be set."

Reply:
Fixed.